File: RiftVibeSolver.Common/Public/Solver.cs
using System;
using System.Collections.Generic;

namespace RiftVibeSolver.Common;

public static class Solver {
    private const double VIBE_LENGTH = 5d;

    public static SolverResult Solve(SolverData data) {
        if (data.HitCount == 0)
            return new SolverResult(0, Array.Empty<Activation>(), Array.Empty<Activation>(), Array.Empty<Activation>(), Array.Empty<Activation>());

        var singleVibeActivationData = GetActivationData(data, 1);
        var doubleVibeActivationData = GetActivationData(data, 2);
        var singleVibeActivations = new Activation[singleVibeActivationData.Count];
        var doubleVibeActivations = new Activation[doubleVibeActivationData.Count];

        for (int i = 0; i < singleVibeActivationData.Count; i++)
            singleVibeActivations[i] = GetActivationFromData(singleVibeActivationData, i, 1);

        for (int i = 0; i < doubleVibeActivationData.Count; i++)
            doubleVibeActivations[i] = GetActivationFromData(doubleVibeActivationData, i, 2);

        var bestNextActivationsByFirstVibeIndex = GetBestNextActivationsByFirstVibeIndex(data, singleVibeActivationData, doubleVibeActivationData, out int totalScore);
        var (bestSingleVibeActivationIndices, bestDoubleVibeActivationIndices) = GetBestActivationIndices(data, singleVibeActivationData, doubleVibeActivationData, bestNextActivationsByFirstVibeIndex);
        var bestSingleVibeActivations = new Activation[bestSingleVibeActivationIndices.Count];
        var bestDoubleVibeActivations = new Activation[bestDoubleVibeActivationIndices.Count];
        int index = 0;

        foreach (int i in bestSingleVibeActivationIndices) {
            bestSingleVibeActivations[index] = singleVibeActivations[i];
            index++;
        }

        index = 0;

        foreach (int i in bestDoubleVibeActivationIndices) {
            bestDoubleVibeActivations[index] = doubleVibeActivations[i];
            index++;
        }

        return new SolverResult(totalScore, singleVibeActivations, doubleVibeActivations, bestSingleVibeActivations, bestDoubleVibeActivations);
    }

    public static VibePath GetVibePath(SolverData data, double startTime, int vibesUsed) {
        var hits = data.Hits;
        var span = GetSpanStartingAt(data, startTime, data.GetFirstHitIndexAfter(startTime), vibesUsed);
        var segments = new List<VibePathSegment>();
        double currentTime = span.StartTime;
        double vibeRemaining = vibesUsed * VIBE_LENGTH;

        for (int i = span.StartIndex; i < span.EndIndex; i++) {
            var hit = hits[i];

            if (!hit.GivesVibe)
                continue;

            segments.Add(new VibePathSegment(currentTime, hit.Time, vibeRemaining, Math.Max(0d, vibeRemaining - (hit.Time - currentTime))));
            vibeRemaining = Math.Max(VIBE_LENGTH, Math.Min(vibeRemaining - (hit.Time - currentTime) + VIBE_LENGTH, 2d * VIBE_LENGTH));
            currentTime = hit.Time;
        }

        segments.Add(new VibePathSegment(currentTime, currentTime + vibeRemaining, vibeRemaining, 0d));
        currentTime += vibeRemaining;

        double maxEndTime = data.GetTimeFromBeat(data.GetBeatFromTime(currentTime) + GetVibeExtensionForBeatLength(data.GetBeatLengthAtTime(currentTime), data.BeatDivisions));
        double endTime = Math.Min(maxEndTime, data.GetHitTime(data.GetFirstHitIndexAfter(currentTime)));

        int score = 0;

        for (int i = span.StartIndex; i < span.EndIndex; i++)
            score += hits[i].Score;

        return new VibePath(startTime, endTime, span.StartIndex, span.EndIndex, score, segments);
    }

    private static List<ActivationData> GetActivationData(SolverData data, int vibesUsed) {
        var activations = new List<ActivationData>();
        int fromIndex = data.GetNextVibe(0);

        if (vibesUsed == 2)
            fromIndex = data.GetNextVibe(fromIndex + 1);

        if (fromIndex == data.HitCount)
            return activations;

        var hits = data.Hits;
        var spans = new List<ActivationSpan>(2 * hits.Count);

        for (int i = fromIndex; i < hits.Count; i++)
            spans.Add(GetSpanStartingAfterHit(data, i, vibesUsed));

        for (int i = fromIndex; i < hits.Count; i++)
            GetSpansEndingOnHit(spans, data, i, vibesUsed);

        GetSpansEndingOnBeats(spans, data, vibesUsed);

        if (spans.Count == 0)
            return activations;

        spans.Sort();

        int currentStartIndex = spans[0].StartIndex;
        int currentEndIndex = spans[0].StartIndex;
        int score = 0;

        foreach (var span in spans) {
            if (span.StartIndex <= fromIndex)
                continue;

            while (currentStartIndex < span.StartIndex) {
                score -= hits[currentStartIndex].Score;
                currentStartIndex++;
            }

            while (currentEndIndex < span.EndIndex) {
                score += hits[currentEndIndex].Score;
                currentEndIndex++;
            }

            while (currentEndIndex > span.EndIndex) {
                currentEndIndex--;
                score -= hits[currentEndIndex].Score;
            }

            if (activations.Count > 0 && activations[activations.Count - 1].MinStartTime == span.StartTime) {
                if (activations[activations.Count - 1].EndIndex >= span.EndIndex)
                    continue;

                activations.RemoveAt(activations.Count - 1);
            }

            activations.Add(new ActivationData(span.StartTime, span.StartIndex, span.EndIndex, score));
        }

        return activations;
    }

    private static ActivationSpan GetSpanStartingAt(SolverData data, double startTime, int firstHitIndex, int vibesUsed) {
        double currentTime = startTime;
        double vibeRemaining = vibesUsed * VIBE_LENGTH;
        double vibeZeroTime = currentTime + vibeRemaining;
        double endTime = GetEndTime();
        int nextVibeIndex = data.GetNextVibe(firstHitIndex);

        while (nextVibeIndex < data.HitCount) {
            double nextVibeTime = data.GetHitTime(nextVibeIndex);

            if (nextVibeTime > endTime)
                break;

            vibeRemaining = Math.Max(VIBE_LENGTH, Math.Min(vibeRemaining - (nextVibeTime - currentTime) + VIBE_LENGTH, 2d * VIBE_LENGTH));
            currentTime = nextVibeTime;
            vibeZeroTime = currentTime + vibeRemaining;
            endTime = GetEndTime();
            nextVibeIndex = data.GetNextVibe(nextVibeIndex + 1);
        }

        return new ActivationSpan(startTime, firstHitIndex, data.GetFirstHitIndexAfter(endTime));

        double GetEndTime() {
            double maxEndTime = data.GetTimeFromBeat(data.GetBeatFromTime(vibeZeroTime) + GetVibeExtensionForBeatLength(data.GetBeatLengthAtTime(vibeZeroTime), data.BeatDivisions));

            return Math.Min(maxEndTime, data.GetHitTime(data.GetFirstHitIndexAfter(vibeZeroTime)));
        }
    }

    private static ActivationSpan GetSpanStartingAfterHit(SolverData data, int hitIndex, int vibesUsed) => GetSpanStartingAt(data, data.GetHitTime(hitIndex), hitIndex + 1, vibesUsed);

    private static void GetSpansEndingOnHit(List<ActivationSpan> spans, SolverData data, int hitIndex, int vibesUsed) {
        double endTime = data.GetHitTime(hitIndex);
        double endBeat = data.GetBeatFromTime(endTime);
        double previousHitTime = data.GetHitTime(hitIndex - 1);
        int endIndex;

        if (data.Hits[hitIndex].GivesVibe)
            endIndex = GetSpanStartingAfterHit(data, hitIndex, 1).EndIndex;
        else
            endIndex = hitIndex + 1;

        if (!data.HasBeatTimings) {
            double vibeZeroTime = data.GetTimeFromBeat(endBeat - GetVibeExtensionForBeatLength(data.GetBeatLengthForBeat((int) endBeat), data.BeatDivisions));

            vibeZeroTime = Math.Max(vibeZeroTime, previousHitTime);
            GetSpansWhereVibeHitsZeroAt(spans, data, vibeZeroTime, endIndex, vibesUsed);

            return;
        }

        for (int i = (int) endBeat; i >= 1; i--) {
            double beatTime = data.GetTimeFromBeat(i);
            double nextBeatTime = data.GetTimeFromBeat(i + 1);
            double vibeZeroTime = data.GetTimeFromBeat(endBeat - GetVibeExtensionForBeatLength(nextBeatTime - beatTime, data.BeatDivisions));

            vibeZeroTime = Math.Max(vibeZeroTime, previousHitTime);

            if (vibeZeroTime > beatTime && vibeZeroTime < nextBeatTime)
                GetSpansWhereVibeHitsZeroAt(spans, data, vibeZeroTime, endIndex, vibesUsed);

            if (beatTime <= previousHitTime)
                break;
        }
    }

    private static void GetSpansEndingOnBeats(List<ActivationSpan> spans, SolverData data, int vibesUsed) {
        var beatTimings = data.BeatTimings;

        if (beatTimings.Count <= 1)
            return;

        for (int i = 1; i < beatTimings.Count - 1; i++) {
            double beatTime = beatTimings[i];
            double endTimeWithPreviousBeatLength = data.GetTimeFromBeat(i + 1 + GetVibeExtensionForBeatLength(beatTime - beatTimings[i - 1], data.BeatDivisions));
            double endTimeWithNextBeatLength = data.GetTimeFromBeat(i + 1 + GetVibeExtensionForBeatLength(beatTimings[i + 1] - beatTime, data.BeatDivisions));
            int nextHitIndex = data.GetFirstHitIndexAfter(beatTime);
            double nextHitTime = data.GetHitTime(nextHitIndex);

            if (!(endTimeWithPreviousBeatLength >= nextHitTime ^ endTimeWithNextBeatLength >= nextHitTime))
                continue;

            int hitIndex = nextHitIndex;

            if (endTimeWithNextBeatLength < nextHitTime)
                hitIndex--;

            int endIndex;

            if (hitIndex < data.HitCount && data.Hits[hitIndex].GivesVibe)
                endIndex = GetSpanStartingAfterHit(data, hitIndex, 1).EndIndex;
            else
                endIndex = hitIndex + 1;

            GetSpansWhereVibeHitsZeroAt(spans, data, beatTime, endIndex, vibesUsed);
        }
    }

    private static void GetSpansWhereVibeHitsZeroAt(List<ActivationSpan> spans, SolverData data, double vibeZeroTime, int endIndex, int vibesUsed) {
        double currentTime = vibeZeroTime;
        double vibeNeeded = 0d;
        int previousVibeIndex = data.GetPreviousVibe(data.GetFirstHitIndexAfter(currentTime));

        do {
            double previousVibeTime = data.GetHitTime(previousVibeIndex);
            double possibleStartTime = currentTime - (vibesUsed * VIBE_LENGTH - vibeNeeded);
            double needFullVibeAt = currentTime - (2d * VIBE_LENGTH - vibeNeeded);

            if (possibleStartTime > previousVibeTime)
                spans.Add(new ActivationSpan(possibleStartTime, data.GetFirstHitIndexAfter(possibleStartTime), endIndex));

            if (previousVibeTime < needFullVibeAt)
                break;

            vibeNeeded = Math.Min(vibeNeeded + (currentTime - previousVibeTime) - VIBE_LENGTH, 2d * VIBE_LENGTH);

            if (vibeNeeded <= 0d)
                break;

            currentTime = previousVibeTime;
            previousVibeIndex = data.GetPreviousVibe(previousVibeIndex);
        } while (previousVibeIndex >= 0);
    }

    private static double GetVibeExtensionForBeatLength(double beatLength, int beatDivisions) {
        double hitWindowInBeats = 0.175d / beatLength;

        return hitWindowInBeats - Math.Floor(hitWindowInBeats * beatDivisions) / beatDivisions;
    }

    private static Dictionary<int, BestNextActivations> GetBestNextActivationsByFirstVibeIndex(SolverData data, List<ActivationData> singleVibeActivations, List<ActivationData> doubleVibeActivations, out int totalScore) {
        var bestNextActivationsByFirstVibeIndex = new Dictionary<int, BestNextActivations>();

        totalScore = GetBestNextActivations(0).BestNextValue;

        return bestNextActivationsByFirstVibeIndex;

        BestNextActivations GetBestNextActivations(int fromIndex) {
            int firstVibeIndex = data.GetNextVibe(fromIndex);

            if (bestNextActivationsByFirstVibeIndex.TryGetValue(firstVibeIndex, out var bestNextActivations))
                return bestNextActivations;

            int secondVibeIndex = data.GetNextVibe(firstVibeIndex + 1);
            int bestNextValue = 0;
            var bestNextSingleVibeActivations = new List<int>();
            var bestNextDoubleVibeActivations = new List<int>();

            for (int i = 0; i < singleVibeActivations.Count; i++) {
                var nextActivation = singleVibeActivations[i];

                if (nextActivation.StartIndex <= firstVibeIndex)
                    continue;

                if (nextActivation.StartIndex > secondVibeIndex)
                    break;

                int nextActivationValue = nextActivation.Score + GetBestNextActivations(nextActivation.EndIndex).BestNextValue;

                if (nextActivationValue > bestNextValue) {
                    bestNextValue = nextActivationValue;
                    bestNextSingleVibeActivations.Clear();
                }

                if (nextActivationValue == bestNextValue)
                    bestNextSingleVibeActivations.Add(i);
            }

            for (int i = 0; i < doubleVibeActivations.Count; i++) {
                var nextActivation = doubleVibeActivations[i];

                if (nextActivation.StartIndex <= secondVibeIndex)
                    continue;

                int nextActivationValue = nextActivation.Score + GetBestNextActivations(nextActivation.EndIndex).BestNextValue;

                if (nextActivationValue > bestNextValue) {
                    bestNextValue = nextActivationValue;
                    bestNextSingleVibeActivations.Clear();
                    bestNextDoubleVibeActivations.Clear();
                }

                if (nextActivationValue == bestNextValue)
                    bestNextDoubleVibeActivations.Add(i);
            }

            bestNextActivations = new BestNextActivations(bestNextValue, bestNextSingleVibeActivations, bestNextDoubleVibeActivations);
            bestNextActivationsByFirstVibeIndex.Add(firstVibeIndex, bestNextActivations);

            return bestNextActivations;
        }
    }

    private static (HashSet<int>, HashSet<int>) GetBestActivationIndices(SolverData data, List<ActivationData> singleVibeActivationData, List<ActivationData> doubleVibeActivationData, Dictionary<int, BestNextActivations> bestNextActivationsByFirstVibeIndex) {
        var bestSingleVibeActivationIndices = new HashSet<int>();
        var bestDoubleVibeActivationIndices = new HashSet<int>();

        Traverse(0);

        return (bestSingleVibeActivationIndices, bestDoubleVibeActivationIndices);

        void Traverse(int fromIndex) {
            var bestNextActivations = bestNextActivationsByFirstVibeIndex[data.GetNextVibe(fromIndex)];

            foreach (int index in bestNextActivations.BestNextSingleVibeActivations) {
                if (bestSingleVibeActivationIndices.Contains(index))
                    continue;

                bestSingleVibeActivationIndices.Add(index);
                Traverse(singleVibeActivationData[index].EndIndex);
            }

            foreach (int index in bestNextActivations.BestNextDoubleVibeActivations) {
                if (bestDoubleVibeActivationIndices.Contains(index))
                    continue;

                bestDoubleVibeActivationIndices.Add(index);
                Traverse(doubleVibeActivationData[index].EndIndex);
            }
        }
    }

    private static Activation GetActivationFromData(List<ActivationData> activationData, int index, int vibesUsed) {
        var activation = activationData[index];

        return new Activation(
            activation.MinStartTime,
            index < activationData.Count - 1 ? activationData[index + 1].MinStartTime : activation.MinStartTime + 1d,
            activation.Score,
            vibesUsed);
    }
}

File: js/solver.js
const Activation = require('./activation');
const ActivationData = require('./activationData');
const ActivationSpan = require('./activationSpan');
const BestNextActivations = require('./bestNextActivations');
const SolverResult = require('./solverResult');
const VibePath = require('./vibePath');
const VibePathSegment = require('./vibePathSegment');

const VIBE_LENGTH = 5.0;

const Solver = {
  solve(data) {
    if (data.hitCount === 0) {
      return new SolverResult(
        0,
        [],
        [],
        [],
        []
      );
    }

    const singleVibeActivationData = getActivationData(data, 1);
    const doubleVibeActivationData = getActivationData(data, 2);
    const singleVibeActivations = singleVibeActivationData.map((_, i) => getActivationFromData(singleVibeActivationData, i, 1));
    const doubleVibeActivations = doubleVibeActivationData.map((_, i) => getActivationFromData(doubleVibeActivationData, i, 2));

    const { bestNextActivationsByFirstVibeIndex, totalScore } = getBestNextActivationsByFirstVibeIndex(
      data,
      singleVibeActivationData,
      doubleVibeActivationData
    );
    const { bestSingle, bestDouble } = getBestActivationIndices(
      data,
      singleVibeActivationData,
      doubleVibeActivationData,
      bestNextActivationsByFirstVibeIndex
    );

    const bestSingleVibeActivations = Array.from(bestSingle).map(i => singleVibeActivations[i]);
    const bestDoubleVibeActivations = Array.from(bestDouble).map(i => doubleVibeActivations[i]);

    return new SolverResult(
      totalScore,
      singleVibeActivations,
      doubleVibeActivations,
      bestSingleVibeActivations,
      bestDoubleVibeActivations
    );
  },

  getVibePath(data, startTime, vibesUsed) {
    const hits = data.hits;
    const span = getSpanStartingAt(data, startTime, data.getFirstHitIndexAfter(startTime), vibesUsed);
    const segments = [];
    let currentTime = span.startTime;
    let vibeRemaining = vibesUsed * VIBE_LENGTH;

    for (let i = span.startIndex; i < span.endIndex; i++) {
      const hit = hits[i];
      if (!hit.givesVibe) continue;
      segments.push(new VibePathSegment(currentTime, hit.time, vibeRemaining, Math.max(0, vibeRemaining - (hit.time - currentTime))));
      vibeRemaining = Math.max(
        VIBE_LENGTH,
        Math.min(vibeRemaining - (hit.time - currentTime) + VIBE_LENGTH, 2 * VIBE_LENGTH)
      );
      currentTime = hit.time;
    }

    segments.push(new VibePathSegment(currentTime, currentTime + vibeRemaining, vibeRemaining, 0));
    currentTime += vibeRemaining;

    const maxEndTime = data.getTimeFromBeat(data.getBeatFromTime(currentTime) + getVibeExtensionForBeatLength(data.getBeatLengthAtTime(currentTime), data.beatDivisions));
    const endTime = Math.min(maxEndTime, data.getHitTime(data.getFirstHitIndexAfter(currentTime)));

    let score = 0;
    for (let i = span.startIndex; i < span.endIndex; i++) score += hits[i].score;

    return new VibePath(startTime, endTime, span.startIndex, span.endIndex, score, segments);
  }
};

function getActivationData(data, vibesUsed) {
  const activations = [];
  let fromIndex = data.getNextVibe(0);
  if (vibesUsed === 2) fromIndex = data.getNextVibe(fromIndex + 1);
  if (fromIndex === data.hitCount) return activations;

  const hits = data.hits;
  const spans = [];

  for (let i = fromIndex; i < hits.length; i++) spans.push(getSpanStartingAfterHit(data, i, vibesUsed));
  for (let i = fromIndex; i < hits.length; i++) getSpansEndingOnHit(spans, data, i, vibesUsed);
  getSpansEndingOnBeats(spans, data, vibesUsed);

  if (spans.length === 0) return activations;

  spans.sort((a, b) => a.startTime - b.startTime);

  let currentStartIndex = spans[0].startIndex;
  let currentEndIndex = spans[0].startIndex;
  let score = 0;

  for (const span of spans) {
    if (span.startIndex <= fromIndex) continue;

    while (currentStartIndex < span.startIndex) {
      score -= hits[currentStartIndex].score;
      currentStartIndex++;
    }
    while (currentEndIndex < span.endIndex) {
      score += hits[currentEndIndex].score;
      currentEndIndex++;
    }
    while (currentEndIndex > span.endIndex) {
      currentEndIndex--;
      score -= hits[currentEndIndex].score;
    }

    if (activations.length > 0 && activations[activations.length - 1].minStartTime === span.startTime) {
      if (activations[activations.length - 1].endIndex >= span.endIndex) continue;
      activations.pop();
    }
    activations.push(new ActivationData(span.startTime, span.startIndex, span.endIndex, score));
  }
  return activations;
}

function getSpanStartingAt(data, startTime, firstHitIndex, vibesUsed) {
  let currentTime = startTime;
  let vibeRemaining = vibesUsed * VIBE_LENGTH;
  let vibeZeroTime = currentTime + vibeRemaining;
  let endTime = getEndTime();
  let nextVibeIndex = data.getNextVibe(firstHitIndex);

  while (nextVibeIndex < data.hitCount) {
    const nextVibeTime = data.getHitTime(nextVibeIndex);
    if (nextVibeTime > endTime) break;
    vibeRemaining = Math.max(
      VIBE_LENGTH,
      Math.min(vibeRemaining - (nextVibeTime - currentTime) + VIBE_LENGTH, 2 * VIBE_LENGTH)
    );
    currentTime = nextVibeTime;
    vibeZeroTime = currentTime + vibeRemaining;
    endTime = getEndTime();
    nextVibeIndex = data.getNextVibe(nextVibeIndex + 1);
  }

  return new ActivationSpan(startTime, firstHitIndex, data.getFirstHitIndexAfter(endTime));

  function getEndTime() {
    const maxEndTime = data.getTimeFromBeat(
      data.getBeatFromTime(vibeZeroTime) +
        getVibeExtensionForBeatLength(data.getBeatLengthAtTime(vibeZeroTime), data.beatDivisions)
    );
    return Math.min(maxEndTime, data.getHitTime(data.getFirstHitIndexAfter(vibeZeroTime)));
  }
}

function getSpanStartingAfterHit(data, hitIndex, vibesUsed) {
  return getSpanStartingAt(data, data.getHitTime(hitIndex), hitIndex + 1, vibesUsed);
}

function getSpansEndingOnHit(spans, data, hitIndex, vibesUsed) {
  const endTime = data.getHitTime(hitIndex);
  const endBeat = data.getBeatFromTime(endTime);
  const previousHitTime = data.getHitTime(hitIndex - 1);
  let endIndex;
  if (data.hits[hitIndex].givesVibe) endIndex = getSpanStartingAfterHit(data, hitIndex, 1).endIndex; else endIndex = hitIndex + 1;

  if (!data.hasBeatTimings) {
    let vibeZeroTime = data.getTimeFromBeat(endBeat - getVibeExtensionForBeatLength(data.getBeatLengthForBeat(Math.floor(endBeat)), data.beatDivisions));
    vibeZeroTime = Math.max(vibeZeroTime, previousHitTime);
    getSpansWhereVibeHitsZeroAt(spans, data, vibeZeroTime, endIndex, vibesUsed);
    return;
  }

  for (let i = Math.floor(endBeat); i >= 1; i--) {
    const beatTime = data.getTimeFromBeatInt(i);
    const nextBeatTime = data.getTimeFromBeatInt(i + 1);
    let vibeZeroTime = data.getTimeFromBeat(
      endBeat - getVibeExtensionForBeatLength(nextBeatTime - beatTime, data.beatDivisions)
    );
    vibeZeroTime = Math.max(vibeZeroTime, previousHitTime);
    if (vibeZeroTime > beatTime && vibeZeroTime < nextBeatTime) {
      getSpansWhereVibeHitsZeroAt(spans, data, vibeZeroTime, endIndex, vibesUsed);
    }
    if (beatTime <= previousHitTime) break;
  }
}

function getSpansEndingOnBeats(spans, data, vibesUsed) {
  const beatTimings = data.beatTimings;
  if (beatTimings.length <= 1) return;
  for (let i = 1; i < beatTimings.length - 1; i++) {
    const beatTime = beatTimings[i];
    const endTimeWithPreviousBeatLength = data.getTimeFromBeat(
      i + 1 + getVibeExtensionForBeatLength(beatTime - beatTimings[i - 1], data.beatDivisions)
    );
    const endTimeWithNextBeatLength = data.getTimeFromBeat(
      i + 1 + getVibeExtensionForBeatLength(beatTimings[i + 1] - beatTime, data.beatDivisions)
    );
    const nextHitIndex = data.getFirstHitIndexAfter(beatTime);
    const nextHitTime = data.getHitTime(nextHitIndex);
    if (!((endTimeWithPreviousBeatLength >= nextHitTime) ^ (endTimeWithNextBeatLength >= nextHitTime))) continue;
    let hitIndex = nextHitIndex;
    if (endTimeWithNextBeatLength < nextHitTime) hitIndex--;
    let endIndex;
    if (hitIndex < data.hitCount && data.hits[hitIndex].givesVibe) {
      endIndex = getSpanStartingAfterHit(data, hitIndex, 1).endIndex;
    } else {
      endIndex = hitIndex + 1;
    }
    getSpansWhereVibeHitsZeroAt(spans, data, beatTime, endIndex, vibesUsed);
  }
}

function getSpansWhereVibeHitsZeroAt(spans, data, vibeZeroTime, endIndex, vibesUsed) {
  let currentTime = vibeZeroTime;
  let vibeNeeded = 0;
  let previousVibeIndex = data.getPreviousVibe(data.getFirstHitIndexAfter(currentTime));

  do {
    const previousVibeTime = data.getHitTime(previousVibeIndex);
    const possibleStartTime = currentTime - (vibesUsed * VIBE_LENGTH - vibeNeeded);
    const needFullVibeAt = currentTime - (2 * VIBE_LENGTH - vibeNeeded);
    if (possibleStartTime > previousVibeTime) {
      spans.push(new ActivationSpan(possibleStartTime, data.getFirstHitIndexAfter(possibleStartTime), endIndex));
    }
    if (previousVibeTime < needFullVibeAt) break;
    vibeNeeded = Math.min(vibeNeeded + (currentTime - previousVibeTime) - VIBE_LENGTH, 2 * VIBE_LENGTH);
    if (vibeNeeded <= 0) break;
    currentTime = previousVibeTime;
    previousVibeIndex = data.getPreviousVibe(previousVibeIndex);
  } while (previousVibeIndex >= 0);
}

function getVibeExtensionForBeatLength(beatLength, beatDivisions) {
  const hitWindowInBeats = 0.175 / beatLength;
  return hitWindowInBeats - Math.floor(hitWindowInBeats * beatDivisions) / beatDivisions;
}

function getBestNextActivationsByFirstVibeIndex(data, singleVibeActivations, doubleVibeActivations) {
  const bestNextActivationsByFirstVibeIndex = new Map();

  const totalScore = getBestNextActivations(0).bestNextValue;
  return { bestNextActivationsByFirstVibeIndex, totalScore };

  function getBestNextActivations(fromIndex) {
    const firstVibeIndex = data.getNextVibe(fromIndex);
    if (bestNextActivationsByFirstVibeIndex.has(firstVibeIndex)) {
      return bestNextActivationsByFirstVibeIndex.get(firstVibeIndex);
    }
    const secondVibeIndex = data.getNextVibe(firstVibeIndex + 1);
    let bestNextValue = 0;
    const bestNextSingleVibeActivations = [];
    const bestNextDoubleVibeActivations = [];

    for (let i = 0; i < singleVibeActivations.length; i++) {
      const nextActivation = singleVibeActivations[i];
      if (nextActivation.startIndex <= firstVibeIndex) continue;
      if (nextActivation.startIndex > secondVibeIndex) break;
      const nextActivationValue = nextActivation.score + getBestNextActivations(nextActivation.endIndex).bestNextValue;
      if (nextActivationValue > bestNextValue) {
        bestNextValue = nextActivationValue;
        bestNextSingleVibeActivations.length = 0;
      }
      if (nextActivationValue === bestNextValue) {
        bestNextSingleVibeActivations.push(i);
      }
    }

    for (let i = 0; i < doubleVibeActivations.length; i++) {
      const nextActivation = doubleVibeActivations[i];
      if (nextActivation.startIndex <= secondVibeIndex) continue;
      const nextActivationValue = nextActivation.score + getBestNextActivations(nextActivation.endIndex).bestNextValue;
      if (nextActivationValue > bestNextValue) {
        bestNextValue = nextActivationValue;
        bestNextSingleVibeActivations.length = 0;
        bestNextDoubleVibeActivations.length = 0;
      }
      if (nextActivationValue === bestNextValue) {
        bestNextDoubleVibeActivations.push(i);
      }
    }

    const best = new BestNextActivations(bestNextValue, bestNextSingleVibeActivations, bestNextDoubleVibeActivations);
    bestNextActivationsByFirstVibeIndex.set(firstVibeIndex, best);
    return best;
  }
}

function getBestActivationIndices(data, singleVibeActivationData, doubleVibeActivationData, bestNextActivationsByFirstVibeIndex) {
  const bestSingle = new Set();
  const bestDouble = new Set();
  traverse(0);
  return { bestSingle, bestDouble };

  function traverse(fromIndex) {
    const bestNext = bestNextActivationsByFirstVibeIndex.get(data.getNextVibe(fromIndex));
    for (const index of bestNext.bestNextSingleVibeActivations) {
      if (!bestSingle.has(index)) {
        bestSingle.add(index);
        traverse(singleVibeActivationData[index].endIndex);
      }
    }
    for (const index of bestNext.bestNextDoubleVibeActivations) {
      if (!bestDouble.has(index)) {
        bestDouble.add(index);
        traverse(doubleVibeActivationData[index].endIndex);
      }
    }
  }
}

function getActivationFromData(activationData, index, vibesUsed) {
  const activation = activationData[index];
  return new Activation(
    activation.minStartTime,
    index < activationData.length - 1 ? activationData[index + 1].minStartTime : activation.minStartTime + 1,
    activation.score,
    vibesUsed
  );
}

module.exports = Solver;
